// mongold/lib/model/ops/write.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

var _index = require('../../index');

var internals = {};

internals.writeAction = function (action) {

  return function () {

    var args = _lodash2['default'].toArray(arguments);
    var collection = this._collection;
    var callback = _assert2['default'].ifError;
    var hasCallback = false;
    var formatResponse;

    if (_lodash2['default'].isFunction(_lodash2['default'].last(args))) {
      callback = args.pop();
      hasCallback = true;
    }

    var document;
    var selector;
    var modifier;
    var options;

    if (action === 'insert' || action === 'save') {
      document = args.shift();

      if (!document || !_lodash2['default'].isObject(document)) {
        return callback(new Error('A document object is required'));
      }
    }

    if (action === 'update' || action === 'remove') {
      selector = args.shift() || {};

      // if (!selector || !_.isObject(selector)) {
      //  return callback(new Error('A selector object is required'));
      // }

      if (selector._id && !(selector._id instanceof _index.Id)) {
        selector._id = new _index.Id(selector._id);
      }

      if (action === 'update') {
        modifier = args.shift();

        if (!modifier || !_lodash2['default'].isObject(modifier)) {
          return callback(new Error('A modifier object is required'));
        }

        // TODO: Maybe validate the modifier?
      }
    }

    options = args.shift() || {};

    _lodash2['default'].defaults(options, {
      validate: true,
      writeConcern: {
        w: hasCallback ? 1 : 0
      }
    });

    if (document && options.validate) {
      document = this.clean(document);

      // Validate the document
      // If there is an error, use the callback
      try {
        this.check(document);
      } catch (error) {
        return callback(error);
      }
    }

    callback = _lodash2['default'].wrap(callback, function (next, error, response) {

      if (error) {
        return next(error);
      }

      if (response.result.ok !== 1) {
        return next(new Error('Database \'' + action + '\' operation failed'));
      }

      next(null, formatResponse(response));
    });

    this.on('ready', function () {

      if (action === 'insert' || action === 'save') {
        // Return the id (or array of ids) as the response
        formatResponse = function (response) {
          // If the document has an id and there are no ops documents, just return the id
          if (document._id && (!response.ops || response.ops.length === 0)) {
            return document._id;
          }

          if (!_lodash2['default'].isArray(document)) {
            return response.ops[0]._id;
          }

          return _lodash2['default'].map(response.ops, function (op) {
            return op._id;
          });
        };

        if (action === 'save') {
          return collection.save(document, options, callback);
        }

        return collection.insert(document, options, callback);
      }

      if (action === 'update' || action === 'remove') {
        // ES6: arrow function
        formatResponse = function (response) {
          return response.result.n;
        };

        // If this is the remove action, fire and don't continue
        if (action === 'remove') {
          _lodash2['default'].defaults(options, { justOne: selector._id ? true : false });
          return collection.remove(selector, callback);
        }

        _lodash2['default'].defaults(options, { multi: selector._id ? false : true });
        return collection.update(selector, modifier, options, callback);
      }
    });
  };
};

var insert = internals.writeAction('insert');
exports.insert = insert;
var save = internals.writeAction('save');
exports.save = save;
var update = internals.writeAction('update');
exports.update = update;
var remove = internals.writeAction('remove');
exports.remove = remove;
